File: tests/localnet/src/pump.ts
/**
 * Real pump instruction builders — remaining accounts per @pump-fun/pump-sdk@1.36.0 pin.
 */
import {
  PublicKey,
  SystemProgram,
  SYSVAR_RENT_PUBKEY,
  TransactionInstruction,
} from "@solana/web3.js";
import {
  ASSOCIATED_TOKEN_PROGRAM_ID,
  TOKEN_PROGRAM_ID,
  getAssociatedTokenAddressSync,
} from "@solana/spl-token";
import {
  MPL_TOKEN_METADATA,
  PUMP_FEE_PROGRAM_ID,
  PUMP_PROGRAM_ID,
} from "./constants.js";

const CREATE_DISC = Buffer.from([24, 30, 200, 40, 5, 28, 7, 119]);
const BUY_EXACT_DISC = Buffer.from([56, 252, 116, 8, 158, 223, 205, 95]);

export function pumpPda(seeds: Buffer[]): PublicKey {
  return PublicKey.findProgramAddressSync(seeds, PUMP_PROGRAM_ID)[0];
}

export function globalPda(): PublicKey {
  return pumpPda([Buffer.from("global")]);
}
export function eventAuthorityPda(): PublicKey {
  return pumpPda([Buffer.from("__event_authority")]);
}
export function mintAuthorityPda(): PublicKey {
  return pumpPda([Buffer.from("mint-authority")]);
}
export function bondingCurvePda(mint: PublicKey): PublicKey {
  return pumpPda([Buffer.from("bonding-curve"), mint.toBuffer()]);
}
export function bondingCurveV2Pda(mint: PublicKey): PublicKey {
  return pumpPda([Buffer.from("bonding-curve-v2"), mint.toBuffer()]);
}
export function creatorVaultPda(creator: PublicKey): PublicKey {
  return pumpPda([Buffer.from("creator-vault"), creator.toBuffer()]);
}
export function globalVolumeAccumulatorPda(): PublicKey {
  return pumpPda([Buffer.from("global_volume_accumulator")]);
}
export function userVolumeAccumulatorPda(user: PublicKey): PublicKey {
  return pumpPda([Buffer.from("user_volume_accumulator"), user.toBuffer()]);
}
export function feeConfigPda(): PublicKey {
  return PublicKey.findProgramAddressSync(
    [Buffer.from("fee_config"), PUMP_PROGRAM_ID.toBuffer()],
    PUMP_FEE_PROGRAM_ID,
  )[0];
}
export function metadataPda(mint: PublicKey): PublicKey {
  return PublicKey.findProgramAddressSync(
    [Buffer.from("metadata"), MPL_TOKEN_METADATA.toBuffer(), mint.toBuffer()],
    MPL_TOKEN_METADATA,
  )[0];
}

function encodeString(s: string): Buffer {
  const b = Buffer.from(s, "utf8");
  const len = Buffer.alloc(4);
  len.writeUInt32LE(b.length);
  return Buffer.concat([len, b]);
}

export function buildCreateIx(params: {
  mint: PublicKey;
  user: PublicKey;
  creator: PublicKey;
  name: string;
  symbol: string;
  uri: string;
}): TransactionInstruction {
  const bondingCurve = bondingCurvePda(params.mint);
  const assocBc = getAssociatedTokenAddressSync(
    params.mint,
    bondingCurve,
    true,
    TOKEN_PROGRAM_ID,
  );
  const data = Buffer.concat([
    CREATE_DISC,
    encodeString(params.name),
    encodeString(params.symbol),
    encodeString(params.uri),
    params.creator.toBuffer(),
  ]);
  return new TransactionInstruction({
    programId: PUMP_PROGRAM_ID,
    keys: [
      { pubkey: params.mint, isSigner: true, isWritable: true },
      { pubkey: mintAuthorityPda(), isSigner: false, isWritable: false },
      { pubkey: bondingCurve, isSigner: false, isWritable: true },
      { pubkey: assocBc, isSigner: false, isWritable: true },
      { pubkey: globalPda(), isSigner: false, isWritable: false },
      { pubkey: MPL_TOKEN_METADATA, isSigner: false, isWritable: false },
      { pubkey: metadataPda(params.mint), isSigner: false, isWritable: true },
      { pubkey: params.user, isSigner: true, isWritable: true },
      { pubkey: SystemProgram.programId, isSigner: false, isWritable: false },
      { pubkey: TOKEN_PROGRAM_ID, isSigner: false, isWritable: false },
      {
        pubkey: ASSOCIATED_TOKEN_PROGRAM_ID,
        isSigner: false,
        isWritable: false,
      },
      { pubkey: SYSVAR_RENT_PUBKEY, isSigner: false, isWritable: false },
      { pubkey: eventAuthorityPda(), isSigner: false, isWritable: false },
      { pubkey: PUMP_PROGRAM_ID, isSigner: false, isWritable: false },
    ],
    data,
  });
}

export function buildBuyExactSolInIx(params: {
  user: PublicKey;
  mint: PublicKey;
  creator: PublicKey;
  spendableSolIn: bigint;
  minTokensOut: bigint;
  feeRecipient: PublicKey;
  buybackFeeRecipient: PublicKey;
}): TransactionInstruction {
  const bondingCurve = bondingCurvePda(params.mint);
  const assocBc = getAssociatedTokenAddressSync(
    params.mint,
    bondingCurve,
    true,
    TOKEN_PROGRAM_ID,
  );
  const associatedUser = getAssociatedTokenAddressSync(
    params.mint,
    params.user,
    true,
    TOKEN_PROGRAM_ID,
  );
  const data = Buffer.alloc(8 + 8 + 8 + 1);
  BUY_EXACT_DISC.copy(data, 0);
  data.writeBigUInt64LE(params.spendableSolIn, 8);
  data.writeBigUInt64LE(params.minTokensOut, 16);
  data[24] = 0;

  return new TransactionInstruction({
    programId: PUMP_PROGRAM_ID,
    keys: [
      { pubkey: globalPda(), isSigner: false, isWritable: false },
      { pubkey: params.feeRecipient, isSigner: false, isWritable: true },
      { pubkey: params.mint, isSigner: false, isWritable: false },
      { pubkey: bondingCurve, isSigner: false, isWritable: true },
      { pubkey: assocBc, isSigner: false, isWritable: true },
      { pubkey: associatedUser, isSigner: false, isWritable: true },
      { pubkey: params.user, isSigner: true, isWritable: true },
      { pubkey: SystemProgram.programId, isSigner: false, isWritable: false },
      { pubkey: TOKEN_PROGRAM_ID, isSigner: false, isWritable: false },
      {
        pubkey: creatorVaultPda(params.creator),
        isSigner: false,
        isWritable: true,
      },
      { pubkey: eventAuthorityPda(), isSigner: false, isWritable: false },
      { pubkey: PUMP_PROGRAM_ID, isSigner: false, isWritable: false },
      {
        pubkey: globalVolumeAccumulatorPda(),
        isSigner: false,
        isWritable: false,
      },
      {
        pubkey: userVolumeAccumulatorPda(params.user),
        isSigner: false,
        isWritable: true,
      },
      { pubkey: feeConfigPda(), isSigner: false, isWritable: false },
      { pubkey: PUMP_FEE_PROGRAM_ID, isSigner: false, isWritable: false },
      {
        pubkey: bondingCurveV2Pda(params.mint),
        isSigner: false,
        isWritable: false,
      },
      {
        pubkey: params.buybackFeeRecipient,
        isSigner: false,
        isWritable: true,
      },
    ],
    data,
  });
}

export { TOKEN_PROGRAM_ID, getAssociatedTokenAddressSync };

File: programs/batchit/src/pump.rs
//! Pump CPI helpers — layout from third_party/pump-sdk/PIN.md (@pump-fun/pump-sdk@1.36.0).

use anchor_lang::prelude::*;
use anchor_lang::solana_program::{
    instruction::{AccountMeta, Instruction},
    program::invoke_signed,
};

pub const PUMP_PROGRAM_ID: Pubkey = pubkey!("6EF8rrecthR5Dkzon8Nwu78hRvfCKubJ14M5uBEwF6P");
pub const PUMP_FEE_PROGRAM_ID: Pubkey = pubkey!("pfeeUxB6jkeY1Hxd7CsFCAjcbHA9rWtchMGdZ6VojVZ");

const BUY_EXACT_SOL_IN_DISCRIMINATOR: [u8; 8] = [56, 252, 116, 8, 158, 223, 205, 95];

/// Bonding curve layout (partial) for reading reserves after buy.
/// Matches BondingCurve in pinned pump-sdk IDL (first fields).
#[derive(Clone, Copy, Debug)]
pub struct BondingCurveReserves {
    pub virtual_token_reserves: u64,
    pub virtual_quote_reserves: u64,
    pub real_token_reserves: u64,
    pub real_quote_reserves: u64,
}

/// Decode reserves from bonding curve account data (8-byte Anchor disc + fields).
pub fn decode_bonding_curve_reserves(data: &[u8]) -> Result<BondingCurveReserves> {
    require!(data.len() >= 8 + 32, crate::error::BatchitError::CreateNotLanded);
    let d = &data[8..];
    Ok(BondingCurveReserves {
        virtual_token_reserves: u64::from_le_bytes(d[0..8].try_into().unwrap()),
        virtual_quote_reserves: u64::from_le_bytes(d[8..16].try_into().unwrap()),
        real_token_reserves: u64::from_le_bytes(d[16..24].try_into().unwrap()),
        real_quote_reserves: u64::from_le_bytes(d[24..32].try_into().unwrap()),
    })
}

pub fn bonding_curve_pda(mint: &Pubkey) -> (Pubkey, u8) {
    Pubkey::find_program_address(&[b"bonding-curve", mint.as_ref()], &PUMP_PROGRAM_ID)
}

pub fn bonding_curve_v2_pda(mint: &Pubkey) -> (Pubkey, u8) {
    Pubkey::find_program_address(&[b"bonding-curve-v2", mint.as_ref()], &PUMP_PROGRAM_ID)
}

pub fn creator_vault_pda(creator: &Pubkey) -> (Pubkey, u8) {
    Pubkey::find_program_address(&[b"creator-vault", creator.as_ref()], &PUMP_PROGRAM_ID)
}

pub fn global_pda() -> (Pubkey, u8) {
    Pubkey::find_program_address(&[b"global"], &PUMP_PROGRAM_ID)
}

pub fn event_authority_pda() -> (Pubkey, u8) {
    Pubkey::find_program_address(&[b"__event_authority"], &PUMP_PROGRAM_ID)
}

pub fn global_volume_accumulator_pda() -> (Pubkey, u8) {
    Pubkey::find_program_address(&[b"global_volume_accumulator"], &PUMP_PROGRAM_ID)
}

pub fn user_volume_accumulator_pda(user: &Pubkey) -> (Pubkey, u8) {
    Pubkey::find_program_address(
        &[b"user_volume_accumulator", user.as_ref()],
        &PUMP_PROGRAM_ID,
    )
}

pub fn fee_config_pda() -> (Pubkey, u8) {
    Pubkey::find_program_address(
        &[b"fee_config", PUMP_PROGRAM_ID.as_ref()],
        &PUMP_FEE_PROGRAM_ID,
    )
}

/// CPI buy_exact_sol_in with buyer PDA as signer.
/// remaining on the outer instruction must already be validated; here we take AccountInfos.
#[allow(clippy::too_many_arguments)]
pub fn cpi_buy_exact_sol_in<'info>(
    buyer: AccountInfo<'info>,
    buyer_seeds: &[&[u8]],
    global: AccountInfo<'info>,
    fee_recipient: AccountInfo<'info>,
    mint: AccountInfo<'info>,
    bonding_curve: AccountInfo<'info>,
    associated_bonding_curve: AccountInfo<'info>,
    associated_user: AccountInfo<'info>,
    creator_vault: AccountInfo<'info>,
    event_authority: AccountInfo<'info>,
    pump_program: AccountInfo<'info>,
    global_volume_accumulator: AccountInfo<'info>,
    user_volume_accumulator: AccountInfo<'info>,
    fee_config: AccountInfo<'info>,
    fee_program: AccountInfo<'info>,
    system_program: AccountInfo<'info>,
    token_program: AccountInfo<'info>,
    bonding_curve_v2: AccountInfo<'info>,
    buyback_fee_recipient: AccountInfo<'info>,
    spendable_sol_in: u64,
    min_tokens_out: u64,
) -> Result<()> {
    let mut data = Vec::with_capacity(25);
    data.extend_from_slice(&BUY_EXACT_SOL_IN_DISCRIMINATOR);
    data.extend_from_slice(&spendable_sol_in.to_le_bytes());
    data.extend_from_slice(&min_tokens_out.to_le_bytes());
    data.push(0); // track_volume = false

    let buyer_key = *buyer.key;
    let accounts = vec![
        AccountMeta::new_readonly(*global.key, false),
        AccountMeta::new(*fee_recipient.key, false),
        AccountMeta::new_readonly(*mint.key, false),
        AccountMeta::new(*bonding_curve.key, false),
        AccountMeta::new(*associated_bonding_curve.key, false),
        AccountMeta::new(*associated_user.key, false),
        AccountMeta::new(buyer_key, true),
        AccountMeta::new_readonly(*system_program.key, false),
        AccountMeta::new_readonly(*token_program.key, false),
        AccountMeta::new(*creator_vault.key, false),
        AccountMeta::new_readonly(*event_authority.key, false),
        AccountMeta::new_readonly(*pump_program.key, false),
        AccountMeta::new_readonly(*global_volume_accumulator.key, false),
        AccountMeta::new(*user_volume_accumulator.key, false),
        AccountMeta::new_readonly(*fee_config.key, false),
        AccountMeta::new_readonly(*fee_program.key, false),
        AccountMeta::new_readonly(*bonding_curve_v2.key, false),
        AccountMeta::new(*buyback_fee_recipient.key, false),
    ];

    let ix = Instruction {
        program_id: PUMP_PROGRAM_ID,
        accounts,
        data,
    };

    let infos = [
        global,
        fee_recipient,
        mint,
        bonding_curve,
        associated_bonding_curve,
        associated_user,
        buyer,
        system_program,
        token_program,
        creator_vault,
        event_authority,
        pump_program,
        global_volume_accumulator,
        user_volume_accumulator,
        fee_config,
        fee_program,
        bonding_curve_v2,
        buyback_fee_recipient,
    ];

    invoke_signed(&ix, &infos, &[buyer_seeds]).map_err(|e| {
        msg!("pump buy CPI failed: {:?}", e);
        error!(crate::error::BatchitError::BuyFailed)
    })?;
    Ok(())
}
